content/posts/replacing-the-redisgraph-parser.md

```diff
@@ -0,0 +1,109 @@
+```console
+MATCH (node1)-->(node2), (src)-[*2..]->(two_or_more_hops)
+```
+To more robust query validations, built-in precedence management for operators and comparisons, and handling for more edge cases than I can recall.
+
+We also took advantage of this moment to migrate all parsing to RedisGraph threads, so multiple queries can be parsed in parallel and the Redis server never blocks to perform this work.
+
+## The strength of strong abstractions
+Beyond these improvements, the greatest benefit of this refactor has been a much stronger abstraction layer between the AST and the tree of operations required to act upon it.
+
+Our earlier grammar made it easy to rely directly on the AST for the construction of operations, especially in superficially simple clauses like CREATE. This approach belies the complexity of circumstances that need to be considered, however. For example, given the query:
+```console
+MATCH (p:Person {name: 'Jeffrey'})
+CREATE (p)-[:EMPLOYED_BY]->(c: Company {name: 'Redis Labs'})
+RETURN *
+```
+The CREATE operation does not exist in isolation; it relies on projections from previous operations and itself  projects data that must be returned to the user.
+
+Having a strong abstraction layer between the AST and the execution tree of operations allows for far easier extension as the complexity of a project increases.
+
+RedisGraph execution plan:
+```console
+MATCH (p:Parser {name: 'libcypher-parser'})-[:CONSTRUCTS]->(ast) RETURN ast
+
+Results
+    Project
+        Conditional Traverse | (p:Parser)->(ast)
+            Filter
+                Node By Label Scan | (p:Parser)
+```
+
+libcypher-parser AST:
+```console
+ @0   0..76  statement                body=@1
+ @1   0..76  > query                  clauses=[@2, @15]
+ @2   0..65  > > MATCH                pattern=@3
+ @3   6..64  > > > pattern            paths=[@4]
+ @4   6..64  > > > > pattern path     (@5)-[@11]-(@13)
+ @5   6..43  > > > > > node pattern   (@6:@7 {@8})
+ @6   7..8   > > > > > > identifier   `p`
+ @7   8..15  > > > > > > label        :`Parser`
+ @8  16..42  > > > > > > map          {@9:@10}
+ @9  17..21  > > > > > > > prop name  `name`
+@10  23..41  > > > > > > > string     "libcypher-parser"
+@11  43..59  > > > > > rel pattern    -[:@12]->
+@12  45..56  > > > > > > rel type     :`CONSTRUCTS`
+@13  59..64  > > > > > node pattern   (@14)
+@14  60..63  > > > > > > identifier   `ast`
+@15  65..76  > > RETURN               projections=[@16]
+@16  72..76  > > > projection         expression=@17
+@17  72..75  > > > > identifier       `ast`
+```
+
+## Lessons learned
+I think this captures the most important things I've learned (outside of my editor window) over the course of this project:
```

Comment:
Change to: "I think this captures the most important lessons I've learned (outside of my editor window) over the course of this project:"